indico/web/client/js/react/components/DateRangePicker.jsx

```diff
@@ -0,0 +1,238 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2024 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import DatePickerCalendar from 'indico/react/components/DatePickerCalendar';
+import {FinalField, validators as v} from 'indico/react/forms';
+import {Param, Translate} from 'indico/react/i18n';
+import {DateRange} from 'indico/utils/date';
+import {formatDate, ISO_FORMAT} from 'indico/utils/date_format';
+import {fromISOLocalDate} from 'indico/utils/date_parser';
+
+import 'indico/custom_elements/ind_date_picker';
+
+const INVALID = '__invalid__';
+
+export default function DateRangePicker({
+  value,
+  label,
+  format,
+  rangeStartLabel,
+  rangeEndLabel,
+  rangeStartMin,
+  rangeStartMax,
+  rangeEndMin,
+  rangeEndMax,
+  min,
+  max,
+  readOnly,
+  disabled,
+  onChange,
+}) {
+  format ??= moment.localeData().longDateFormat('L');
+
+  if (min) {
+    rangeStartMin = rangeStartMax = min;
+  }
+
+  if (max) {
+    rangeEndMin = rangeEndMax = max;
+  }
+
+  function handleChange(ev) {
+    const picker = ev.currentTarget;
+    const [start, end] = picker.dateRange;
+    const rangeStartInput = picker.querySelector('[data-range-start]');
+    const rangeEndInput = picker.querySelector('[data-range-end]');
+
+    const startInvalid = rangeStartInput.value && !start;
+    const endInvalid = rangeEndInput.value && !end;
+
+    onChange({
+      // ISO local dates
+      startDate: startInvalid ? INVALID : formatDate(ISO_FORMAT, start) || null,
+      endDate: endInvalid ? INVALID : formatDate(ISO_FORMAT, end) || null,
+    });
+  }
+
+  return (
+    <ind-date-range-picker
+      range-start={fromISOLocalDate(value.startDate)?.toDateString() || ''}
+      range-end={fromISOLocalDate(value.endDate)?.toDateString() || ''}
+      range-start-min={rangeStartMin}
+      range-start-max={rangeStartMax}
+      range-end-min={rangeEndMin}
+      range-end-max={rangeEndMax}
+      onChange={handleChange}
+    >
+      <fieldset>
+        <legend>{label}</legend>
+        <label>
+          <span>{rangeStartLabel}:</span>
+          <input
+            type="text"
+            data-range-start
+            placeholder={format}
+            readOnly={readOnly}
+            disabled={disabled}
```

Comment:
Actually, there's another feature which we need to keep: The ability to lock the start date. When you edit a booking that had its first occurrence in the past, you cannot change its start date (since we don't want to "change history") anymore:

![image](https://github.com/user-attachments/assets/fc04f9a2-ed41-4dac-ba20-3476fadc4d56)

Reply:
Hmmm, that kinda sounds like you want to use a single date picker with a text label that says what the start date is.

Reply:
I thought about that for a moment as well, but you don't see the range that way when opening the picker. It's also more straightforward to just have a prop to specify this since - at least in case of editing bookings - it depends on when the booking started. So by being able to disable the start date it's more consistent than having to make a single date picker look like date range picker (even though I guess the way you did the components it would not be too hard, since the same CSS Can be used :))

Reply:
Ah ok, yeah, it's nicer to see the range, true. I think we can work this in relatively easily if we change the behavior as I mentioned [in the other comment](https://github.com/indico/indico/pull/6464#discussion_r1734482951). I could make the start date readonly, and then not allow it to be changed in that case.